### ISSUE TITLE
Deleting a file while a command is running crashes Rome

### ISSUE THREAD

**Issue by ooflorent** [romejs/rome#389]
## Steps To Reproduce

1. Create an empty file `dummy.ts`
2. Start a long running command such as `rome lint`
3. Delete `dummy.ts` while the command is running

Doing this would cause Rome to crash:

```
rome master % ./scripts/dev-rome lint --save
 Building trunk
ℹ Bundling packages/@romejs/cli/bin/rome.ts
✔ Saved to /var/folders/gp/z7cw34td69v0qlsytclcm9600000gn/T/rome-dev
 Executing trunk
⚠ Disk caching has been disabled due to the ROME_CACHE=0 environment variable
 Linting: packages/@romejs/js-compiler/scope/evaluators/Program.ts                                 elapsed 1s eta 14s 153/1_336[Error: Expected file stats for /Users/fcailhol/Development/rome/packages/@romejs/parser-core/lol.ts
  at ___R$project$rome$$romejs$core$master$fs$MemoryFileSystem_ts$default.getFileStatsAssert (/private/var/folders/gp/z7cw34td69v0qlsytclcm9600000gn/T/rome-dev/index.js:100500:14) generated source location
  at ___R$project$rome$$romejs$core$master$fs$FileAllocator_ts$default.handleDeleted (/private/var/folders/gp/z7cw34td69v0qlsytclcm9600000gn/T/rome-dev/index.js:98708:41) generated source location
  at process.processTicksAndRejections (internal/process/task_queues.js:97:4)]
✖ Fatal error occurred: Error: Expected file stats for /Users/fcailhol/Development/rome/packages/@romejs/parser-core/lol.ts
    at ___R$project$rome$$romejs$core$master$fs$MemoryFileSystem_ts$default.getFileStatsAssert
  (/private/var/folders/gp/z7cw34td69v0qlsytclcm9600000gn/T/rome-dev/index.js:100500:14) generated source location
    at ___R$project$rome$$romejs$core$master$fs$FileAllocator_ts$default.handleDeleted
  (/private/var/folders/gp/z7cw34td69v0qlsytclcm9600000gn/T/rome-dev/index.js:98708:41) generated source location
    at process.processTicksAndRejections (internal/process/task_queues.js:97:4)
```

## Expected Results

Rome should not crash.

**Comment by sebmck** [comment]
Looks like a race condition.

**Comment by yassere** [comment]
MemoryFileSystem#handleDeletion calls `deletedFileEvent.send(path)` after it calls `files.delete(path)`

In response to the `deletedFileEvent`, FileAllocator#handleDeleted calls `memoryFs.getFileStatsAssert(path)` which checks the `files` Map for the path entry which was just deleted, and then throws an error when it can't find it.

I think you can solve this particular problem by changing MemoryFileSystem#handleDeletion to an async function then awaiting `this.deletedFileEvent.call(path)` before deleting the `files` entry.

But I don't know if that would cause other problems, and rome still crashes on file deletion for other reasons that I haven't dug into yet.

**Comment by sebmck** [comment]
Ah yeah changing it to `await .call()` sounds right. There may be some issues such as duplicate fs watch events that could cause race conditions but those could be solved with locks.

**Comment by yassere** [comment]
`LintRunner` populates worker queues with all the paths that were passed to `LintRunner#run` at the start of the run.

There doesn't appear to be any handling of file deletion while those queues spin down, and when `Cache#get` is called for a nonexistent path, rome crashes because `Cache#createEmptyEntry` needs the `mtime`.

**Comment by sebmck** [comment]
Hopefully resolved #604. Let's reopen this if there's more issues.
